### ISSUE TITLE
[SvelteKit] Error: Cannot use `cookies.set(...)` after the response has been generated

### ISSUE THREAD

**Issue by chanmathew** [supabase/auth-helpers#466]
# Bug report

## Describe the bug

When using SvelteKit's new [streaming promises feature](https://kit.svelte.dev/docs/load#streaming-with-promises), it seems to run into an issue setting cookies with the logic that runs in [server hooks mentioned here](https://supabase.com/docs/guides/auth/auth-helpers/sveltekit#migration-set-up-hooks).

Here is the error:
```
Error: Cannot use `cookies.set(...)` after the response has been generated
    at Object.event.cookies.set (/node_modules/@sveltejs/kit/src/runtime/server/respond.js:422:11)
    at fetch (/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:150:21)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.eval [as fetch] (/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:32:10)
    at async generateItinerary (/src/routes/(app)/trips/[tripId]/itinerary/+page.server.ts:7:22)
```

If I comment out all the logic in my hooks file, the error seems to go away, and the data from the streamed promise will return correctly.

Here is my `+page.server.ts` load function that is calling the API endpoint:

```
export const load = (async ({ parent, fetch, locals, params }) => {
	const generateItinerary = async () => {
		const response = await fetch('/api/itineraries/generate', {
			method: 'POST',
			body: JSON.stringify({
				destination: 'Paris, France',
				days: 1,
				preferences: 'must-see attractions'
			})
		});

		const data = await response.json(); <<< This is where it breaks
		console.log('DATA: ', data); <<< This console log never happens due to the cookie error thrown

		return data;
	};

	return {
		lazy: {
			itineraries: generateItinerary()
		}
	};
}) satisfies PageServerLoad;
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Setup `hooks.server.ts` as instructed according to Supabase docs
2. Create an API endpoint
3. Create a layout or page server load function to fetch from API endpoint
4. The load function will fail to return the data from the API endpoint, and error appears server console

## Expected behavior

The load function should unwrap the streaming promise from the API endpoint and return the data.

## System information

- OS: macOS
- Version of supabase-js: v2.10.0
- Version of supabase-auth-helpers-sveltekit: v0.9.0
- Version of Node.js: v16.17.0
- Version of SvelteKit: v1.10.0

**Comment by david-plugge** [comment]
Thanks for the report.

The problem here is that sveltekit generates the response when all load functions finished. At this moment the headers are sent, as this has to happen before we can send the response body.
Therefore, this does not only affect auth helpers, but all situations where you want to set a cookie this way.
To work around this limitation you can implement it in a way like this:

_src/routes/+page.server.ts_
```ts
import type { PageServerLoad } from './$types';

export const load = (async ({ fetch, locals: { getSession } }) => {
	// get the session before load finishes
	const session = await getSession();

	const loadPosts = async () => {
		const response = await fetch('/api/posts', {
			headers: {
				Authorization: `Bearer ${session?.access_token}`
			}
		});

		const data = await response.json();
		return data.posts;
	};

	return {
		lazy: {
			posts: loadPosts()
		}
	};
}) satisfies PageServerLoad;
```

_src/routes/api/posts/+server.ts_
```ts
import { PUBLIC_SUPABASE_ANON_KEY, PUBLIC_SUPABASE_URL } from '$env/static/public';
import { createClient } from '@supabase/supabase-js';
import { json } from '@sveltejs/kit';
import type { RequestHandler } from './$types';

export const GET: RequestHandler = async ({ request }) => {
	const [_, access_token] = request.headers.get('Authorization')?.split(' ') ?? [];

	const supabase = createClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
		auth: { autoRefreshToken: false, persistSession: false }
	});
	supabase.auth.setSession({ access_token, refresh_token: '' });

	const { data: posts, error } = await supabase.from('posts').select();
	console.log({ posts, error });

	return json({
		posts
	});
};
```

If you want to lazily load data from supabase without an extra endpoint, you can do it like this:

```ts
// create a new supabase client with the access token
// this way we don´t have to rely on cookies
function createLazyClient(accessToken: string = '') {
	const supabase = createClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
		auth: {
			autoRefreshToken: false,
			persistSession: false
		}
	});
	supabase.auth.setSession({
		access_token: accessToken,
		refresh_token: ''
	});

	return supabase;
}

export const load = (async ({ locals: { getSession } }) => {
	const session = await getSession();
	const lazySupabase = createLazyClient(session?.access_token);

	const loadPosts = async () => {
		// simulate delay
		await new Promise((res) => setTimeout(res, 1000));

		const { data: posts } = await lazySupabase.from('posts').select();

		return posts;
	};

	return {
		lazy: {
			posts: loadPosts()
		}
	};
}) satisfies PageServerLoad;
```

**Comment by chanmathew** [comment]
Thanks so much @david-plugge for your response.

I've tried the 2nd method by initiating a lazySupabase client and then making a streaming promise call within one of my load functions, but unfortunately I still seem to be running into the same issue.

Do I need to also change every other instance of Supabase across my load function, as well as in my hooks.server.ts?

**Comment by david-plugge** [comment]
> Do I need to also change every other instance of Supabase across all my load functions, as well as in my hooks.server.ts?

No, the helpers should work fine when used outside of lazy loaded promises. 
Could you share your code? It´s easier to check for errors with an example.

**Comment by chanmathew** [comment]
@david-plugge 

```ts
import { PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY } from '$env/static/public';
import { createClient } from '@supabase/supabase-js';
import { error } from '@sveltejs/kit';
import type { LayoutServerLoad } from './$types';

function createLazyClient(accessToken = '') {
	const supabase = createClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
		auth: {
			autoRefreshToken: false,
			persistSession: false
		}
	});
	supabase.auth.setSession({
		access_token: accessToken,
		refresh_token: ''
	});

	return supabase;
}

export const load = (async ({ params, locals, fetch }) => {
	const session = await locals.getSession();
	const lazySupabase = createLazyClient(session?.access_token);

	const fetchItineraries = async () => {
		const { data, error: err } = await lazySupabase
			.from('trips_itineraries')
			.select('*')
			.eq('trips_id', params.tripId);

		if (err) {
			throw error(500, err.message);
		}

		if (!data?.length) {
			const response = await fetch('/api/itineraries/generate', {
				method: 'POST',
				body: JSON.stringify({
					destination: 'Paris, France',
					days: 1,
					preferences: 'must-see attractions'
				})
			});

			const { choices } = await response.json();
			const itineraries = JSON.parse(choices[0].message.content);

			return itineraries;
		}

		return data;
	};

	return {
		lazy: {
			itineraries: fetchItineraries()
		}
	};
}) satisfies LayoutServerLoad;
```

**Comment by david-plugge** [comment]
Where exactly does the error appear? You are fetching some data from a sveltekit api endpoint, are you using supabase in there or is this fetch call not even executed?

**Comment by chanmathew** [comment]
@david-plugge It's not using Supabase in that API endpoint, and it does execute, however it breaks when it's trying to return the response from the API endpoint back to the load function:

This is essentially what my API endpoint is doing:
```ts
import type { RequestHandler } from './$types';
import { error, type Config } from '@sveltejs/kit';

export const config: Config = {
	runtime: 'edge'
};

export const POST: RequestHandler = async ({ request, fetch }) => {
	try {
		const requestData = await request.json();

		const response = await fetch('https://someendpoint', {
			headers: {
				Authorization: `Bearer ${OPENAI_API_KEY}`,
				'Content-Type': 'application/json'
			},
			method: 'POST',
			body: JSON.stringify(requestData)
		});

		const data = await response.json();

		return new Response(JSON.stringify(data), {
			status: 200,
			statusText: 'Itinerary generated successfully.'
		});
	} catch (err) {
		console.error(err);
		throw error(500, 'An error occurred');
	}
};

```

This is the error I get:

```
src/routes/(app)/trips/[tripId]/+layout.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
/node_modules/@sveltejs/kit/src/runtime/server/respond.js:422
				throw new Error('Cannot use `cookies.set(...)` after the response has been generated');
				      ^

Error: Cannot use `cookies.set(...)` after the response has been generated
    at Object.event.cookies.set (/node_modules/@sveltejs/kit/src/runtime/server/respond.js:422:11)
    at fetch (/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:150:21)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.eval [as fetch] (/node_modules/@sveltejs/kit/src/runtime/server/fetch.js:32:10)
    at async fetchItineraries (/src/routes/(app)/trips/[tripId]/+layout.server.ts:31:24)
```

**Comment by david-plugge** [comment]
ouhh thats interesting. So what happens if you don´t use fetch from load but rather the global fetch with the full url (http://localhost:5173/api/itineraries/generate)

**Comment by david-plugge** [comment]
Thinking about this a bit more it may also be caused by the request scoped supabase instance. This is getting kind of tricky....

**Comment by chanmathew** [comment]
> ouhh thats interesting. So what happens if you don´t use fetch from load but rather the global fetch with the full url (http://localhost:5173/api/itineraries/generate)
> 
> And also, can you check if the api returns a `set-cookie` header? Sveltekit would try to merge them into the actual request which could result in the error beeing thrown.

Oh this is interesting, when I use a global fetch function with the full URL the set cookie error goes away.

Here's the headers from the response from the API endpoint:
<img width="511" alt="CleanShot 2023-03-10 at 16 15 22@2x" src="https://user-images.githubusercontent.com/2940761/224429812-aaf60d5d-693a-4a42-8350-50b187165b17.png">

**Comment by david-plugge** [comment]
Yeah so fetch from load directly forwards the request to the api endpoint if you pass in a url without an origin, completely skipping the http layer. It also tries to emulate the clients behaviour by sending the cookies along. Could you try passing `credentials: 'omit'` to the load fetch? Not sure if they implemented that though.

**Comment by chanmathew** [comment]
@david-plugge oh that worked! Here's the headers:
<img width="660" alt="CleanShot 2023-03-10 at 16 22 20@2x" src="https://user-images.githubusercontent.com/2940761/224430885-24648bec-7373-49d4-8680-c9d0ea4f6d2d.png">

**Comment by david-plugge** [comment]
Glad to hear!

**Comment by chanmathew** [comment]
Hi @david-plugge - thanks for your help again with this, just wanted to raise this as I realized there are still issues with the current approach of omitting credentials. For example, if I'm protecting my API routes following the [guide here](https://supabase.com/docs/guides/auth/auth-helpers/sveltekit#protecting-api-routes) or [here](https://supabase.com/docs/guides/auth/auth-helpers/sveltekit#protecting-multiple-routes): 

And I have this in my hooks.server.ts
```ts
	// protect API requests to all routes that start with /api
	if (event.url.pathname.startsWith('/api')) {
		const session = await event.locals.getSession();

		if (!session) {
			// the user is not signed in
			throw error(401, { message: 'Unauthorized' });
		}
	}
```

The `session` in that case is always null, and therefore will not be able to hit the endpoint since it'll throw a 401.

Do you think there would be any workarounds in the meantime?

And to clarify, would this actually be an issue the auth-helpers maintainers would be figuring out a patch for or is this more on the SvelteKit team?

**Comment by david-plugge** [comment]
I wonder if there is an actual benefit of using this approach (streaming). It might be easier to just load the _lazy_ data on the client so you don´t have to worry about all this trouble.

But if you still prefer streaming data you would need to create a custom supabase client, similar to `createLazyClient` if the request has a `Authorization` header. This isn´t tested at all, just written of the top of my head.

```ts
export const handle: Handle = ({ event, resolve }) => {

	if (event.request.headers.has('Authorization')) {
		const [_, token] = event.request.headers.get('Authorization').split(' ');

		if (token) {
			event.locals.supabase = createLazyClient(token);
		}
	}

	if (!event.locals.supabase) {
		// default supabase auth helpers implementation
	}

	event.locals.getSession = async () => {
		const {
			data: { session }
		} = await event.locals.supabase.auth.getSession();
		return session;
	};

	return resolve(event, {
		/**
		 * There´s an issue with `filterSerializedResponseHeaders` not working when using `sequence`
		 *
		 * https://github.com/sveltejs/kit/issues/8061
		 */
		filterSerializedResponseHeaders(name) {
			return name === 'content-range';
		}
	});
}
```

This client is not able to modify the actual session at all!

**Comment by JowinJ** [comment]
Hi, 

In case this helps anyone else out that lands here look for a solution to this error, I was also having this issue but when trying to log a user in. It turned out to be I had forgotten to await .signInWithPassword() so it was trying to set the cookie after the response had been sent. Oops

**Comment by tobiassern** [comment]
I am able to reproduce this error with the example "Sveltekit Email password" with no changes. It is easier to reproduce when setting "JWT expiry limit" to 30 seconds.

Then navigate to the base path ("/") via the link in the layout.

From my testing it seems to be some issue with using load function in +page.ts file and having actions in +page.server.ts. When removing the +page.server.ts the error doesn't appear.

https://github.com/supabase/auth-helpers/tree/main/examples/sveltekit-email-password

**Comment by david-plugge** [comment]
@tobiassern can you reproduce it with an expiry limit of 90 seconds? The server client always refreshes the session when it´s valid for less than 60 seconds by default.

**Comment by tobiassern** [comment]
@david-plugge Thanks for the information, didn't know that.

But I can still reproduce it with an expiry limit of 90 seconds. Although if I move the load function to page.server.ts and removes page.ts it works as intended.

**Comment by rudgalvis** [comment]
[Minimal reproduction](https://github.com/rudgalvis/MR-cookie-set-after-response-error)

I have created a minimal, reproducible example for this error and carefully documented it in the README file. Hopefully, it serves useful in resolving the issue, as it's currently blocking our production release.

**Comment by david-plugge** [comment]
Thank you very much for the reproduction @rudgalvis !

I´ve been playing around with it and it seems like the supabase client isn´t actually used when navigating between `/posts` and `/auth`. SvelteKit makes a request to the server and since there is a shared layout load function at the root that does all the work the response is basicly empty (just some sveltekit internal stuff telling the client it has all data).
So what happens is that the client will be created in the handle hook and lazily calls `getSession` some time in the future (no idea why tbh). I managed to create a workaround:

```diff
+let called = false;
 event.locals.getSession = async () => {
+    called = true;
     const {
         data: { session }
     } = await event.locals.supabase.auth.getSession();
     return session;
 };
 
+if (!called) {
+    await event.locals.supabase.auth.initialize();
+}
```

This makes sure `getSession` is called before the response is created (the headers are sent).

I´m not sure how we should actually make this work properly without this weird workaround, it feels like there is  something missing to sveltekit like a simple boolean to check if the headers are already sent. I´ll think about it over the weekend but for the meantime the code snippet above should help you getting rid of the error.

**Comment by mennankara** [comment]
> ```diff
> +let called = false;
>  event.locals.getSession = async () => {
> +    called = true;
>      const {
>          data: { session }
>      } = await event.locals.supabase.auth.getSession();
>      return session;
>  };
>  
> +if (!called) {
> +    await event.locals.getSession();
> +}
> ```
> 
> This makes sure `getSession` is called before the response is created (the headers are sent).
> 

This didn't work, instance suffered from same problem and crashed. Using the exact code from https://supabase.com/docs/guides/getting-started/tutorials/with-sveltekit

**Comment by leerobert** [comment]
This seems to still be an issue with versions:

```
		"@supabase/auth-helpers-sveltekit": "^0.10.3",
		"@supabase/supabase-js": "^2.38.0",
		"@sveltejs/adapter-auto": "^2.0.0",
		"@sveltejs/kit": "^1.20.4",
```

**Comment by jordvisser** [comment]
I've the same happening, when a user opens a url in a blank browser window and their sessions token needs a refresh, reproduced it by setting the refresh timeout to a low number -> got a sessions-> closed the window -> waited -> navigated to the url in blank window after the set timeout was expired.

I've been able to resolve hard crashes, on my occurrence of this error, by adding `try-catch` blocks to the event.cookies.set & event.cookies.delete functions with the following in the `hooks.server.ts` file:
```
    event.locals.supabase = createServerClient(
        PUBLIC_SUPABASE_URL,
        PUBLIC_SUPABASE_ANON_KEY,
        {
            cookies: {
                get: (key) => event.cookies.get(key),
                set: (key, value, options) => {
                    try {
                        event.cookies.set(key, value, options);   
                    } catch (err) {
                        console.error({err})
                    }
                },
                remove: (key, options) => {
                    try {
                        event.cookies.delete(key, options);  
                    } catch (err) {
                        console.error({err})
                    }
                }
            }
        }
    );
```

This is far from ideal, the needed cookie actions do not get applied.

My interpretation of the error, in my setup, is that `createServerClient` is using an async method to set the cookies, but these resolve/execute after sveltekit already passed on the `event` to the `resolve` function.

See the sveltekit source where the `event.cookie.set` function gets overloaded: [sveltejs/kit/.../runtime/server/respond.js](https://github.com/sveltejs/kit/blob/dd1e306ff4e351d1a6b542c7d1f1d6a1181a659a/packages/kit/src/runtime/server/respond.js#L501-L508)
_This is the only occurrence in could find where this error text is used in all of svelte's repositories_
```
......
		} finally {
			event.cookies.set = () => {
				throw new Error('Cannot use `cookies.set(...)` after the response has been generated');
			};

			event.setHeaders = () => {
				throw new Error('Cannot use `setHeaders(...)` after the response has been generated');
			};
		}
......
```

I have to add that my async/await skills are mediocre at best, my analysis might be way off.

**Comment by jordvisser** [comment]
I came to a new 'solution', to make sure the `createServerClient` is done with all the setting/removing of cookies I await a call to `...supabase.auth.getSession()` before returning the `resolve` function in the `hooks.server.ts` `handle`.
_I removed the `try-catch` blocks as they might hide other severe errors._

The code I added to `hooks.server.ts`:
```
const { error: getSessionError } = await event.locals.supabase.auth.getSession()
if (getSessionError) {
    console.error(getSessionError)
}
```

The code whole `hooks.server.ts` as it is right now:
```
import {
    PUBLIC_SUPABASE_ANON_KEY,
    PUBLIC_SUPABASE_URL
} from '$env/static/public';
import { createServerClient } from '@supabase/ssr';
import type { Handle } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {
    event.locals.supabase = createServerClient(
        PUBLIC_SUPABASE_URL,
        PUBLIC_SUPABASE_ANON_KEY,
        {
            cookies: {
                get: (key) => event.cookies.get(key),
                set: (key, value, options) => {
                    event.cookies.set(key, value, options);
                },
                remove: (key, options) => {
                    event.cookies.delete(key, options);
                }
            }
        }
    );

    event.locals.getSession = async () => {
        const {
            data: { session }
        } = await event.locals.supabase.auth.getSession();
        return session;
    };
    

    const { error: getSessionError } = await event.locals.supabase.auth.getSession()
    if (getSessionError) {
        console.error(getSessionError)
    }

    return resolve(event, {
        filterSerializedResponseHeaders(name) {
            return name === 'content-range';
        }
    });
};
```

For now I'm logging any errors that get passed back by `getSession()`, is that a good practice?

Before settling on awaiting `getSession()` I tried awaiting `...supabase.auth.initializePromise`, but this did not prevent the error from occurring.
I would be nice to have something that indicates if `createServerClient` is still busy, with which we can await or do something else to make sure no cookies/headers get set after the `event` is passed on to `resolve`.
_The code snippet I used to try to use the `initializePromise`_
```
const {error: supabaseAuthInitializeError} = await event.locals.supabase.auth.initializePromise;
if (supabaseAuthInitializeError) {
    console.error(supabaseAuthInitializeError);
}
```

**Comment by igorlanko** [comment]
I'm facing this error when trying to make a call to `/api/endpoint/+server.ts` from a companion Chrome extension. My hook intercepts the request and attaches the bearer token to initiate the supabase client with. Other than that the code is the same as from the supabase docs.

I tried awaiting for the session as some suggested here, but the result seems to be the same. SvelteKit is at `1.29.1`. 

Does anyone know if this is a thing with the SSR package? Should I switch?

**Comment by chanmathew** [comment]
I'm still running into this issue when I updated to the SSR package. Here's my versions:
"@sveltejs/kit": "^2.0.0",
"svelte": "5.0.0-next.55",
"supabase": "^1.145.4",
"@supabase/ssr": "^0.1.0",
"@supabase/supabase-js": "^2.39.2",
